File: circuits/circuits/cert_verifier.ts
import * as fs from 'fs';
import * as util from 'util';
import { exec } from 'child_process';
const execAsync = util.promisify(exec);

//Convert .crt certificate file to type .pem for openssl


//Verify the corresponding public key of the certificate used to sign the passport is a valid DS certificate public key
const pubCheck = execAsync('openssl x509 -modulus -noout -in myserver.crt | openssl md5');

//Verify the corresponding public key of the certificate used to sign the DS certificate is a valid CSCA certificate
const DSCheck = execAsync('openssl verify -CAfile CSCA_Certificate.crt -untrusted DS_Certificate.crt');

File: circuits/circuits/cert_verifier.py
import json
from subprocess import check_output
import codecs
import sys
#pass the passport json filepath as passportJson to python function
def main(passportJson):
    f = open(passportJson, "r")
    data = json.load(f)[0]
    modulus = hex(int(data["modulus"]))[2:].upper()

    #verify that the CSCA and DS certificates share the same modulus, ie are chained
    decrypted_CSCA = codecs.decode(check_output(['openssl', 'x509', '-modulus', '-noout', '-in', '../inputs/certificates/CSCA_Certificate.crt']))[8:-2]
    decrypted_DS = codecs.decode(check_output(['openssl', 'x509', '-modulus', '-noout', '-in', '../inputs/certificates/DS_Certificate.crt']))[8:-2]
    if decrypted_CSCA != decrypted_DS or decrypted_DS != modulus:
        raise Exception("F Certificates do not match")

#To run this:
#   String[] command = {"python", "cert_verifier.py", "passport data filepath",}; 
#   Process p = Runtime.getRuntime().exec(command); 
main(sys.argv[1])

#If you want to test, just run this in the terminal: 
#   python3 cert_verifier.py ../inputs/passport_verification/public_key.json
#note that the argument is the passport data filepath.
